src/lsp/ProjectManager.spec.ts

```diff
@@ -341,6 +341,183 @@ describe('ProjectManager', () => {
+            expect(
+                manager.projects.map(x => x.projectKey).sort()
+            ).to.eql([
+                s`${rootDir}/bsconfig.json`,
+                s`${rootDir}/level1/bsconfig.json`,
+                s`${rootDir}/level1/level2/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/level4/bsconfig.json`
+            ]);
+        });
+
+        it('respects maxDepth of 20 when discovering projects', async () => {
+            // Create bsconfig.json files at different depths
+            fsExtra.outputFileSync(`${rootDir}/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/level6/bsconfig.json`, '');
+
+            await manager.syncProjects([{
+                ...workspaceSettings,
+                languageServer: {
+                    ...workspaceSettings.languageServer,
+                    maxDepth: 20
+                }
+            }]);
+
+            // maxDepth: 20 should find all the files we created (since 20 > 6)
+            expect(
+                manager.projects.map(x => x.projectKey).sort()
+            ).to.eql([
+                s`${rootDir}/bsconfig.json`,
+                s`${rootDir}/level1/bsconfig.json`,
+                s`${rootDir}/level1/level2/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/level4/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/level4/level5/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/level4/level5/level6/bsconfig.json`
+            ]);
+        });
+
+        it('uses default maxDepth of 15 when no maxDepth is specified', async () => {
+            // Create bsconfig.json files at different depths
+            fsExtra.outputFileSync(`${rootDir}/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/level6/bsconfig.json`, '');
```

Comment:
Prove that it stops at level 15. Ensure that a file at 15is found, and 16is not found. But you don't need to make all 15 files....you can skip some steps in between.